### ISSUE TITLE
neuronx-cc compile crash `unexpected operator <ufunc 'invert'>`

### ISSUE THREAD

**Issue by cmrfrd** [aws-neuron/aws-neuron-sdk#685]
I am implementing dropout in XLA HLO but get an unexpected compiler error.

This is the debug ir representation:

```
pre-compilation debug_ir: HloModule xla_computation_ordered_wrapper, entry_computation_layout={()->(f32[1000]{0}, f32[1000]{0})}

ENTRY main.55 () -> (f32[1000], f32[1000]) {
  constant.1 = f32[] constant(1)
  broadcast.2 = f32[1000]{0} broadcast(constant.1), dimensions={}
  iota.7 = u32[1000]{0} iota(), iota_dimension=0
  constant.5 = u32[] constant(12345678)
  broadcast.6 = u32[1000]{0} broadcast(constant.5), dimensions={}
  not.9 = u32[1000]{0} not(broadcast.6)
  and.10 = u32[1000]{0} and(iota.7, not.9)
  not.8 = u32[1000]{0} not(iota.7)
  and.11 = u32[1000]{0} and(not.8, broadcast.6)
  or.12 = u32[1000]{0} or(and.10, and.11)
  constant.13 = u32[] constant(16)
  broadcast.14 = u32[1000]{0} broadcast(constant.13), dimensions={}
  shift-right-logical.15 = u32[1000]{0} shift-right-logical(or.12, broadcast.14)
  not.17 = u32[1000]{0} not(shift-right-logical.15)
  and.18 = u32[1000]{0} and(or.12, not.17)
  not.16 = u32[1000]{0} not(or.12)
  and.19 = u32[1000]{0} and(not.16, shift-right-logical.15)
  or.20 = u32[1000]{0} or(and.18, and.19)
  constant.21 = u32[] constant(2246822507)
  broadcast.22 = u32[1000]{0} broadcast(constant.21), dimensions={}
  multiply.23 = u32[1000]{0} multiply(or.20, broadcast.22)
  constant.24 = u32[] constant(13)
  broadcast.25 = u32[1000]{0} broadcast(constant.24), dimensions={}
  shift-right-logical.26 = u32[1000]{0} shift-right-logical(multiply.23, broadcast.25)
  not.28 = u32[1000]{0} not(shift-right-logical.26)
  and.29 = u32[1000]{0} and(multiply.23, not.28)
  not.27 = u32[1000]{0} not(multiply.23)
  and.30 = u32[1000]{0} and(not.27, shift-right-logical.26)
  or.31 = u32[1000]{0} or(and.29, and.30)
  constant.32 = u32[] constant(3266489909)
  broadcast.33 = u32[1000]{0} broadcast(constant.32), dimensions={}
  multiply.34 = u32[1000]{0} multiply(or.31, broadcast.33)
  constant.35 = u32[] constant(16)
  broadcast.36 = u32[1000]{0} broadcast(constant.35), dimensions={}
  shift-right-logical.37 = u32[1000]{0} shift-right-logical(multiply.34, broadcast.36)
  not.39 = u32[1000]{0} not(shift-right-logical.37)
  and.40 = u32[1000]{0} and(multiply.34, not.39)
  not.38 = u32[1000]{0} not(multiply.34)
  and.41 = u32[1000]{0} and(not.38, shift-right-logical.37)
  or.42 = u32[1000]{0} or(and.40, and.41)
  constant.43 = u32[] constant(1)
  broadcast.44 = u32[1000]{0} broadcast(constant.43), dimensions={}
  multiply.45 = u32[1000]{0} multiply(or.42, broadcast.44)
  constant.3 = u32[] constant(1417339264)
  broadcast.4 = u32[1000]{0} broadcast(constant.3), dimensions={}
  compare.46 = pred[1000]{0} compare(multiply.45, broadcast.4), direction=GE
  constant.49 = f32[] constant(1.49253738)
  broadcast.50 = f32[1000]{0} broadcast(constant.49), dimensions={}
  constant.47 = f32[] constant(0)
  broadcast.48 = f32[1000]{0} broadcast(constant.47), dimensions={}
  select.51 = f32[1000]{0} select(compare.46, broadcast.50, broadcast.48)
  multiply.52 = f32[1000]{0} multiply(broadcast.2, select.51)
  ROOT tuple.53 = (f32[1000]{0}, f32[1000]{0}) tuple(multiply.52, select.51)
}
```

In code:

```rust
fn dropout(
    x: &Tensor,
    dropout_probability: f32,
    seed: u32,
    training: bool,
    graph: &mut Graph,
) -> Tensor {
    assert!(dropout_probability >= 0.0 && dropout_probability <= 1.0);

    let split = Tensor::<u32>::full(
        ((dropout_probability.max(0.0).min(1.0)) * u32::MAX as f32) as u32,
        &x.shape(),
        false,
        graph,
    );
    let seed_tensor = Tensor::<u32>::full(seed, &x.shape(), false, graph);
    let num_elements = x.shape().iter().product::<u64>();
    let iota_tensor: Tensor<u32> = iota(&[num_elements], 0, graph, false);
    let iota_tensor_reshaped = reshape(&iota_tensor, &x.shape(), graph);
    let r1: Tensor<u32> = xor(&iota_tensor_reshaped, &seed_tensor, graph);

    // r2 = ((r1 >> 16) ^ r1) * 0x85ebca6b;
    let r2 = &mul(
        &xor(
            &r1,
            &shift_right_logical(
                &r1,
                &Tensor::<u32>::full(16, &x.shape(), false, graph),
                graph,
            ),
            graph,
        ),
        &Tensor::<u32>::full(0x85ebca6b, &x.shape(), false, graph),
        graph,
    );

    // r3 = ((r2 >> 13) ^ r2) * 0xc2b2ae35;
    let r3 = &mul(
        &xor(
            &r2,
            &shift_right_logical(
                &r2,
                &Tensor::<u32>::full(13, &x.shape(), false, graph),
                graph,
            ),
            graph,
        ),
        &Tensor::<u32>::full(0xc2b2ae35, &x.shape(), false, graph),
        graph,
    );

    // r4 = (r3 >> 16) ^ r3;
    let r4 = &mul(
        &xor(
            &r3,
            &shift_right_logical(
                &r3,
                &Tensor::<u32>::full(16, &x.shape(), false, graph),
                graph,
            ),
            graph,
        ),
        &Tensor::<u32>::ones(&x.shape(), false, graph),
        graph,
    );

    let bit_mask = ge(&r4, &split, graph);
    let zeros = Tensor::zeros(&x.shape(), false, graph);
    let keep_prob_value = 1. / (1. - dropout_probability);
    let keep_prob = Tensor::full(D::from_float(keep_prob_value), &x.shape(), false, graph);
    let mask = select(&bit_mask, &keep_prob, &zeros, graph);

    mul(x, &mask, graph)
}
```

This reults in the compiler error:

```
Compiling NEFF /tmp/dropout_seeded_434234159073558259.neff...
Running command 'CUDA_VISIBLE_DEVICES=0 XLA_PYTHON_CLIENT_PREALLOCATE=false neuronx-cc compile /tmp/rust_hlo_dropout_seeded.pb --framework XLA --target trn1 --model-type transformer --auto-cast none --output /tmp/dropout_seeded_434234159073558259.neff'
thread 'nn::tests::dropout_seeded' panicked at 'Script 'CUDA_VISIBLE_DEVICES="0" XLA_PYTHON_CLIENT_PREALLOCATE="false" "neuronx-cc" "compile" "/tmp/rust_hlo_dropout_seeded.pb" "--framework" "XLA" "--target" "trn1" "--model-type" "transformer" "--auto-cast" "none" "--output" "/tmp/dropout_seeded_434234159073558259.neff"' failed: 
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]: ***************************************************************
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]:  An Internal Compiler Error has occurred
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]: ***************************************************************
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]: 
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]: Error message:  unexpected operator <ufunc 'invert'>
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]: 
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]: Error class:    AssertionError
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]: Error location: Unknown
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]: Command line:   /home/ubuntu/.local/bin/neuronx-cc compile /tmp/rust_hlo_dropout_seeded.pb --framework XLA --target trn1 --model-type transformer --auto-cast none --output /tmp/dropout_seeded_434234159073558259.neff
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]: 
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]: Internal details:
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]:   File "neuronxcc/driver/CommandDriver.py", line 259, in neuronxcc.driver.CommandDriver.CommandDriver.run
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]:   File "neuronxcc/driver/commands/CompileCommand.py", line 1089, in neuronxcc.driver.commands.CompileCommand.CompileCommand.run
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]:   File "neuronxcc/driver/commands/CompileCommand.py", line 1040, in neuronxcc.driver.commands.CompileCommand.CompileCommand.runPipeline
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]:   File "neuronxcc/driver/commands/CompileCommand.py", line 1065, in neuronxcc.driver.commands.CompileCommand.CompileCommand.runPipeline
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]:   File "neuronxcc/driver/commands/CompileCommand.py", line 1069, in neuronxcc.driver.commands.CompileCommand.CompileCommand.runPipeline
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]:   File "neuronxcc/driver/Job.py", line 300, in neuronxcc.driver.Job.SingleInputJob.run
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]:   File "neuronxcc/driver/Job.py", line 326, in neuronxcc.driver.Job.SingleInputJob.runOnState
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]:   File "neuronxcc/driver/Pipeline.py", line 30, in neuronxcc.driver.Pipeline.Pipeline.runSingleInput
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]:   File "neuronxcc/driver/Job.py", line 300, in neuronxcc.driver.Job.SingleInputJob.run
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]:   File "neuronxcc/driver/Job.py", line 326, in neuronxcc.driver.Job.SingleInputJob.runOnState
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]:   File "neuronxcc/driver/jobs/Frontend.py", line 359, in neuronxcc.driver.jobs.Frontend.Frontend.runSingleInput
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]:   File "neuronxcc/driver/jobs/Frontend.py", line 164, in neuronxcc.driver.jobs.Frontend.Frontend.runXLAFrontend
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]:   File "neuronxcc/starfish/penguin/Penguin.py", line 299, in neuronxcc.starfish.penguin.Penguin.runPenguin
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]:   File "neuronxcc/starfish/penguin/Frontend.py", line 150, in neuronxcc.starfish.penguin.Frontend.tensorizeXla
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]:   File "neuronxcc/starfish/penguin/Frontend.py", line 151, in neuronxcc.starfish.penguin.Frontend.tensorizeXla
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]:   File "neuronxcc/starfish/penguin/Frontend.py", line 159, in neuronxcc.starfish.penguin.Frontend.tensorizeXla
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]:   File "neuronxcc/starfish/penguin/Frontend.py", line 206, in neuronxcc.starfish.penguin.Frontend.tensorizeXlaFromFile
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]:   File "neuronxcc/starfish/penguin/ModuleGen.py", line 128, in neuronxcc.starfish.penguin.ModuleGen.load_xla_function
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]:   File "neuronxcc/starfish/penguin/ModuleGen.py", line 125, in neuronxcc.starfish.penguin.ModuleGen.load_xla_function.read
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]:   File "<string>", line 90, in <module>
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]:   File "neuronxcc/starfish/penguin/ir/Operator.py", line 463, in neuronxcc.starfish.penguin.ir.Operator.UnaryOp.__init__
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]:   File "neuronxcc/starfish/penguin/ir/Operator.py", line 205, in neuronxcc.starfish.penguin.ir.Operator.Operator.__init__
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]: 
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]: Version information:
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]:   NeuronX Compiler version 2.6.0.19+3d819e565
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]:   
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]:   Python version 3.8.10
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]:   HWM version 2.6.0.0-826e77395
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]:   NEFF version Dynamic
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]:   TVM not available
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]:   NumPy version 1.22.4
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]:   MXNet not available
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]: 
2023-06-10T18:52:46Z ERROR 291220 [neuronx-cc]: Artifacts stored in: /home/ubuntu/Kholinar/xla/neuronxcc-o8l49y1f
```
Here is a .pb: [rust_hlo_dropout_seeded.pb.zip](https://github.com/aws-neuron/aws-neuron-sdk/files/11713094/rust_hlo_dropout_seeded.pb.zip)

I am using `iota` in the computation so could this be related to #659 ?

**Comment by xanderdunn** [comment]
This is work we're doing for deterministic dropout (#676)

The unique thing about this error is the `unexpected operator <ufunc 'invert'>`, I haven't seen this before.

**Comment by hannanjgaws** [comment]
Thank you for reaching out! We are taking a look at this issue, and will provide an update on the ticket once we have a root cause.

**Comment by xanderdunn** [comment]
I updated to Neuron SDK 2.11 and still unable to compile this graph, although the error appears to be different.

Example Graph: [rust_hlo_dropout_seeded.pb.zip](https://github.com/aws-neuron/aws-neuron-sdk/files/11776838/rust_hlo_dropout_seeded.pb.zip)

Compile:
```
$ neuronx-cc compile /tmp/rust_hlo_dropout_seeded.pb --framework XLA --target trn1 --model-type transformer --auto-cast none --output /tmp/dropout_seeded_9121989869275399962.neff
python3.8: /opt/brazil-pkg-cache/packages/LLVM/LLVM-11.x.867.0/AL2_x86_64/DEV.STD.PTHREAD/build/include/llvm/ADT/StringSwitch.h:189: llvm::StringSwitch<T, R>::operator R() [with T = bir::ActivationFunctionType; R = bir::ActivationFunctionType]: Assertion `Result && "Fell off the end of a string-switch"' failed.
Aborted (core dumped)
```

Versions:
```
$ dpkg -l | grep -i neuron
ii  aws-neuronx-collectives                2.14.9.0-f3942f597                amd64        neuron_ccom built using CMake
ii  aws-neuronx-dkms                       2.10.11.0                         amd64        aws-neuronx driver in DKMS format.
ii  aws-neuronx-oci-hook                   2.2.9.0                           amd64        neuron_oci_hook built using CMake
ii  aws-neuronx-runtime-lib                2.14.8.0-edaf1b171                amd64        neuron_runtime built using CMake
ii  aws-neuronx-tools                      2.11.10.0                         amd64        Neuron profile and debug tools

$ neuronx-cc --version
NeuronX Compiler version 2.7.0.40+f7c6cf2a3

Running on AMI ami-0f05159f0fd60f5e1
Running in region usw2-az4
Running on python 3.8.10

HWM version 2.7.0.3-0092b9d34
NumPy version 1.21.6
```

**Comment by aws-rhsoln** [comment]
We are able to reproduce the issue. The issue happens because we currently don’t support invert aka bitwise_not. After carefully looking at the graph, it looks like the xor lowering uses a not operator which resulted in this failure. We have a fix for this issue and it would be available in the upcoming release.

**Comment by xanderdunn** [comment]
Note that this is fixed in neuronx-cc 2.8 beta:
```
$ neuronx-cc --version
NeuronX Compiler version 2.8.0.13+a76a82727

Running on AMI ami-01257e71ecb2f431c
Running in region usw2-az4
Running on python 3.8.10

HWM version 2.8.0.3-2b7c6da39
NumPy version 1.21.6
```

Thanks for the fix! @cmrfrd or AWS, you can close this issue.
